modules/gdscript/gdscript.cpp

```diff
@@ -1387,33 +1387,105 @@ String GDScript::debug_get_script_name(const Ref<Script> &p_script) {
 #endif
 
 thread_local GDScript::UpdatableFuncPtr GDScript::func_ptrs_to_update_thread_local;
+GDScript::UpdatableFuncPtr *GDScript::func_ptrs_to_update_main_thread = &func_ptrs_to_update_thread_local;
 
-GDScript::UpdatableFuncPtrElement GDScript::_add_func_ptr_to_update(GDScriptFunction **p_func_ptr_ptr) {
-	UpdatableFuncPtrElement result = {};
+GDScript::UpdatableFuncPtrElement *GDScript::_add_func_ptr_to_update(GDScriptFunction **p_func_ptr_ptr) {
+	MutexLock lock(func_ptrs_to_update_mutex);
+
+	List<UpdatableFuncPtrElement>::Element *result = func_ptrs_to_update_elems.push_back(UpdatableFuncPtrElement());
 
 	{
-		MutexLock lock(func_ptrs_to_update_thread_local.mutex);
-		result.element = func_ptrs_to_update_thread_local.ptrs.push_back(p_func_ptr_ptr);
-		result.mutex = &func_ptrs_to_update_thread_local.mutex;
+		MutexLock lock2(func_ptrs_to_update_thread_local.mutex);
+		result->get().element = func_ptrs_to_update_thread_local.ptrs.push_back(p_func_ptr_ptr);
+		result->get().mutex = &func_ptrs_to_update_thread_local.mutex;
 
 		if (likely(func_ptrs_to_update_thread_local.initialized)) {
-			return result;
+			return &result->get();
 		}
 
 		func_ptrs_to_update_thread_local.initialized = true;
 	}
 
-	MutexLock lock(func_ptrs_to_update_mutex);
 	func_ptrs_to_update.push_back(&func_ptrs_to_update_thread_local);
 
-	return result;
+	return &result->get();
 }
 
-void GDScript::_remove_func_ptr_to_update(const UpdatableFuncPtrElement p_func_ptr_element) {
-	ERR_FAIL_NULL(p_func_ptr_element.element);
-	ERR_FAIL_NULL(p_func_ptr_element.mutex);
-	MutexLock lock(*p_func_ptr_element.mutex);
-	p_func_ptr_element.element->erase();
+void GDScript::_remove_func_ptr_to_update(const UpdatableFuncPtrElement *p_func_ptr_element) {
+	ERR_FAIL_NULL(p_func_ptr_element);
+	ERR_FAIL_NULL(p_func_ptr_element->element);
+	ERR_FAIL_NULL(p_func_ptr_element->mutex);
+	MutexLock lock(*p_func_ptr_element->mutex);
+	p_func_ptr_element->element->erase();
+}
```

Comment:
Is it okay to lock the mutex after the null checks? What if `p_func_ptr_element` exists when you null check it, but another thread frees it by the time you lock the mutex?

Reply:
First of all, I think none of the checks can fail unless there's a bug. Neither `element` nor `mutex` are ever nullified (if they were nullified here to mark the deletion or something, the lock should happen before indeed). Regarding the whole `p_func_ptr_element`, its users (`GDScriptLambda*Callable`) should never try to call this function with it being null. Because of that, I've added an `if` to `GDScriptLambdaSelfCallable`, which can indeed have it as null. I think that's cleaner. I won't remove the checks now, for extra safety, though, but in the end my advise would be to remove them. We could keep a `DEV_ASSERT(p_func_ptr_element != nullptr)`, but a null pointer there would be very easy to spot while debugging a crash anyway.

For completeness, let me add that multiple theards can't call this function on the same `UpdatableFuncPtrElement`. There's one of those per lambda, and a lambda is destroyed on one thread. If multiple threads were trying to run its destructor, something would be very broken.

Reply:
Would it be more appropriate to make the function take a `const UpdateableFuncPtrElement &` instead of a `const UpdateableFuncPtrElement *` so the type checker can ensure the parameter isn't null?

Reply:
It may be a good idea. However, I stayed away from doing so since that would propagate back to the caller, which right now would need to dereference what is already a pointer for it, with the ugly `*something` notation. Something worth doing in an upcoming iteration.